src/main/java/com/google/cloud/flink/bigquery/BigQuerySourceFunction.java

```diff
@@ -67,77 +71,70 @@ public TypeInformation<RowData> getProducedType() {
+    deserializer.open(() -> getRuntimeContext().getMetricGroup().addGroup("bigQuery"));
+    this.executerIndex = getRuntimeContext().getIndexOfThisSubtask();
     this.numOfExecutors = getRuntimeContext().getNumberOfParallelSubtasks();
     this.numOfStreams = readSessionStreamList.size();
-    for (int i = executorIndex; i < numOfStreams; i += numOfExecutors) {
+    this.streamNames.clear();
+    for (int i = executerIndex; i < numOfStreams; i += numOfExecutors) {
       if (running) {
         this.streamNames.add(readSessionStreamList.get(i));
       }
     }
   }
 
+  @SuppressWarnings("resource")
   @Override
-  public void run(final SourceContext<RowData> ctx) throws Exception {
-    Collector<RowData> collector = new SourceContextCollector<RowData>(ctx);
+  public void run(SourceContext<RowData> ctx) throws Exception {
+    List<RowData> outputCollector = new ArrayList<>();
+    ListCollector<RowData> listCollector = new ListCollector<>(outputCollector);
     Options options =
         new ReadRowsHelper.Options(
-            /* maxRetries= */ 5,
+            /* maxReadRowsRetries= */ 5,
             Optional.of("endpoint"),
             /* backgroundParsingThreads= */ 5,
-            /* prebufferResponses= */ 1);
+            1);
+    if (!streamNames.isEmpty()) {
+      for (String streamName : streamNames) {
+        ReadRowsRequest.Builder readRowsRequest =
+            ReadRowsRequest.newBuilder().setReadStream(streamName);
+        ReadRowsHelper readRowsHelper =
+            new ReadRowsHelper(bigQueryReadClientFactory, readRowsRequest, options);
 
-    for (String streamName : streamNames) {
-      ReadRowsRequest.Builder readRowsRequest =
-          ReadRowsRequest.newBuilder().setReadStream(streamName);
-      try (ReadRowsHelper readRowsHelper =
-          new ReadRowsHelper(bigQueryReadClientFactory, readRowsRequest, options)) {
         Iterator<ReadRowsResponse> readRows = readRowsHelper.readRows();
         while (readRows.hasNext()) {
           ReadRowsResponse response = readRows.next();
           try {
             if (response.hasArrowRecordBatch()) {
               Preconditions.checkState(response.hasArrowRecordBatch());
-              deserializer.deserialize(
-                  response.getArrowRecordBatch().getSerializedRecordBatch().toByteArray(),
-                  collector);
+              deserializer.deserialize(response.toByteArray(), (Collector<RowData>) listCollector);
+
             } else if (response.hasAvroRows()) {
               Preconditions.checkState(response.hasAvroRows());
-              deserializer.deserialize(
-                  response.getAvroRows().getSerializedBinaryRows().toByteArray(), collector);
+              deserializer.deserialize(response.toByteArray(), (Collector<RowData>) listCollector);
               break;
             }
           } catch (IOException ex) {
-            log.error("Error while deserialization", ex);
+            log.error("Error while deserialization");
```

Comment:
Why the exception was removed from the log?

Reply:
Added exception back in the logger